The job **`i686-mingw-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/4914130687/jobs/8775092255) [(plain)](https://github.com/rust-lang-ci/rust/commit/390e41af383ca43944ee332c68e3f3cb1e84b122/checks/13309477386/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args panics\default-backtrace-ice.rs`

error: 1 errors occurred comparing output.
status: exit code: 101
command: PATH="C:\a\rust\rust\build\i686-pc-windows-gnu\stage2\bin;C:\a\rust\rust\build\i686-pc-windows-gnu\stage0-bootstrap-tools\i686-pc-windows-gnu\release\deps;C:\a\rust\rust\build\i686-pc-windows-gnu\stage0\bin;C:\a\rust\rust\ninja;C:\a\rust\rust\mingw32\bin;C:\hostedtoolcache\windows\Python\3.11.3\x64\Scripts;C:\hostedtoolcache\windows\Python\3.11.3\x64;C:\msys64\usr\bin;C:\a\rust\rust\sccache;C:\Program Files\MongoDB\Server\5.0\bin;C:\aliyun-cli;C:\vcpkg;C:\cf-cli;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\2.9.3\x64;C:\cabal\bin;C:\ghcup\bin;C:\Program Files\dotnet;C:\mysql\bin;C:\Program Files\R\R-4.3.0\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\Program Files (x86)\sbt\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\npm\prefix;C:\hostedtoolcache\windows\go\1.20.3\x64\bin;C:\hostedtoolcache\windows\Python\3.7.9\x64\Scripts;C:\hostedtoolcache\windows\Python\3.7.9\x64;C:\hostedtoolcache\windows\Ruby\2.5.9\x64\bin;C:\Program Files\OpenSSL\bin;C:\tools\kotlinc\bin;C:\hostedtoolcache\windows\Java_Temurin-Hotspot_jdk\8.0.372-7\x64\bin;C:\Program Files\ImageMagick-7.1.1-Q16-HDRI;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\Program Files\Eclipse Foundation\jdk-8.0.302.8-hotspot\bin;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\ProgramData\Chocolatey\bin;C:\Program Files\PowerShell\7;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\Program Files\TortoiseSVN\bin;C:\Program Files\CMake\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.7\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\nodejs;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\Program Files\GitHub CLI;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\SeleniumWebDrivers\ChromeDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files (x86)\Google\Cloud SDK\google-cloud-sdk\bin;C:\Program Files (x86)\Microsoft BizTalk Server;C:\Program Files\LLVM\bin;C:\Users\runneradmin\.dotnet\tools;C:\Users\runneradmin\.cargo\bin;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2\\bin\\rustc.exe" "C:\\a\\rust\\rust\\tests\\ui\\panics\\default-backtrace-ice.rs" "-Zthreads=1" "--sysroot" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\stage2" "--target=i686-pc-windows-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=C:\\a\\rust\\rust\\tests\\ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\ui\\panics\\default-backtrace-ice" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\native\\rust-test-helpers" "-L" "C:\\a\\rust\\rust\\build\\i686-pc-windows-gnu\\test\\ui\\panics\\default-backtrace-ice\\auxiliary" "-Z" "treat-err-as-bug=1"
stdout: none
error[E0425]: cannot find value `missing_ident` in this scope
error[E0425]: cannot find value `missing_ident` in this scope
  --> fake-test-src-base\panics\default-backtrace-ice.rs:21:13
   |
LL | fn main() { missing_ident; }


thread 'rustc' panicked at 'aborting due to `-Z treat-err-as-bug=1`', compiler\rustc_errors\src\lib.rs:1711:30
stack backtrace:
   0: 0x66f1e3fe - core::fmt::write::hf5380ac3ee3d71aa
   1: 0x66eb4184 - std::io::Write::write_fmt::hc306171dc7ebfc12
   2: 0x66ebd990 - std::sys_common::backtrace::print::h57d6ebb4a714f582
   3: 0x66ec078f - std::panicking::default_hook::{{closure}}::h942d09aeaff11c5f
   4:  0x1256358 - <unknown>
   5: 0x66ebfdb6 - ___rdl_dealloc
   6: 0x66ec0c68 - std::panicking::begin_panic_handler::{{closure}}::h2330c75e4d60f760
   7: 0x6a43bb07 - <rustc_errors[2007af0c6b9b7a02]::HandlerInner>::panic_if_treat_err_as_bug
   8: 0x6a43b10e - <rustc_errors[2007af0c6b9b7a02]::HandlerInner>::emit_diagnostic::{closure#2}
   9: 0x67597ad3 - rustc_interface[3670b4b1319935bc]::callbacks::track_diagnostic
  10: 0x6a43a09f - <rustc_errors[2007af0c6b9b7a02]::HandlerInner>::emit_diagnostic
  11: 0x6a4440d8 - <rustc_span[1fc5373ceb26c49]::ErrorGuaranteed as rustc_errors[2007af0c6b9b7a02]::diagnostic_builder::EmissionGuarantee>::diagnostic_builder_emit_producing_guarantee
  12: 0x684cc8fe - <rustc_resolve[8f76d2cca51f3187]::Resolver>::report_errors
  13: 0x68490f77 - <rustc_session[7463595a7fe00c8]::session::Session>::time::<(), <rustc_resolve[8f76d2cca51f3187]::Resolver>::resolve_crate::{closure#0}>
  14: 0x684f53b1 - <rustc_resolve[8f76d2cca51f3187]::Resolver>::resolve_crate
  15: 0x6754f328 - rustc_interface[3670b4b1319935bc]::passes::resolver_for_lowering
  16: 0x691b411b - rustc_query_system[d5db14f87540071f]::query::plumbing::try_execute_query::<rustc_query_impl[c5e22ac7ce001d03]::queries::resolver_for_lowering, rustc_query_impl[c5e22ac7ce001d03]::plumbing::QueryCtxt>
  17: 0x68f989bf - rustc_query_impl[c5e22ac7ce001d03]::get_query::resolver_for_lowering
  18: 0x6746d3d2 - <rustc_middle[cf9c5a59c6c6eca]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[1203a84557310c48]::run_compiler::{closure#1}::{closure#2}::{closure#2}, &rustc_data_structures[be249329499d1260]::steal::Steal<(rustc_middle[cf9c5a59c6c6eca]::ty::ResolverAstLowering, alloc[6cb854f4b384d7a6]::rc::Rc<rustc_ast[e3ff4c0dfdf8e34c]::ast::Crate>)>>
  19: 0x674a867d - <rustc_interface[3670b4b1319935bc]::queries::QueryResult<&rustc_middle[cf9c5a59c6c6eca]::ty::context::GlobalCtxt>>::enter::<&rustc_data_structures[be249329499d1260]::steal::Steal<(rustc_middle[cf9c5a59c6c6eca]::ty::ResolverAstLowering, alloc[6cb854f4b384d7a6]::rc::Rc<rustc_ast[e3ff4c0dfdf8e34c]::ast::Crate>)>, rustc_driver_impl[1203a84557310c48]::run_compiler::{closure#1}::{closure#2}::{closure#2}>
  20: 0x674a968a - <rustc_interface[3670b4b1319935bc]::interface::Compiler>::enter::<rustc_driver_impl[1203a84557310c48]::run_compiler::{closure#1}::{closure#2}, core[11d4c4719e5f176b]::result::Result<core[11d4c4719e5f176b]::option::Option<rustc_interface[3670b4b1319935bc]::queries::Linker>, rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>>
  21: 0x67473d73 - std[72a8a6c0d59fa6f5]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3670b4b1319935bc]::util::run_in_thread_pool_with_globals<rustc_interface[3670b4b1319935bc]::interface::run_compiler<core[11d4c4719e5f176b]::result::Result<(), rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>, rustc_driver_impl[1203a84557310c48]::run_compiler::{closure#1}>::{closure#0}, core[11d4c4719e5f176b]::result::Result<(), rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[11d4c4719e5f176b]::result::Result<(), rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>>
  22: 0x6746653a - <<std[72a8a6c0d59fa6f5]::thread::Builder>::spawn_unchecked_<rustc_interface[3670b4b1319935bc]::util::run_in_thread_pool_with_globals<rustc_interface[3670b4b1319935bc]::interface::run_compiler<core[11d4c4719e5f176b]::result::Result<(), rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>, rustc_driver_impl[1203a84557310c48]::run_compiler::{closure#1}>::{closure#0}, core[11d4c4719e5f176b]::result::Result<(), rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[11d4c4719e5f176b]::result::Result<(), rustc_span[1fc5373ceb26c49]::ErrorGuaranteed>>::{closure#1} as core[11d4c4719e5f176b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  23: 0x66ed3569 - __ZN3std3sys7windows6thread6Thread3new12thread_start17h2507f15c1a234d1eE@4
  24: 0x774105b9 - <unknown>
  25: 0x77e077ad - <unknown>
  26: 0x77e0777d - <unknown>
error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md


note: rustc 1.71.0-nightly (390e41af3 2023-05-08) running on i686-pc-windows-gnu

note: compiler flags: -Z threads=1 -C codegen-units=1 -Z ui-testing -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z deduplicate-diagnostics=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 -Z treat-err-as-bug=1
query stack during panic:
#0 [resolver_for_lowering] getting the resolver for lowering
end of query stack
------------------------------------------
---
test result: FAILED. 14713 passed; 1 failed; 234 ignored; 0 measured; 0 filtered out; finished in 944.46s

Some tests failed in compiletest suite=ui mode=ui host=i686-pc-windows-gnu target=i686-pc-windows-gnu
Build completed unsuccessfully in 0:57:47
make: *** [Makefile:80: ci-mingw-subset-2] Error 1
```

</details>

--- a/compiler/rustc_mir_dataflow/src/value_analysis.rs
+++ b/compiler/rustc_mir_dataflow/src/value_analysis.rs
@@ -675,6 +675,15 @@ impl Map {
                 self.cache_preorder_invoke(place);
             }
         }
+
+        // Trim useless places.
+        for opt_place in self.locals.iter_mut() {
+            if let Some(place) = *opt_place && self.inner_values[place].is_empty() {
+                *opt_place = None;
+            }
+        }
+        #[allow(rustc::potential_query_instability)]
+        self.projections.retain(|_, child| !self.inner_values[*child].is_empty());
     }
 
     /// Potentially register the (local, projection) place and its fields, recursively.
@@ -803,7 +812,7 @@ impl Map {
         tail_elem: Option<TrackElem>,
         f: &mut impl FnMut(ValueIndex),
     ) {
-        if place.is_indirect() {
+        if place.has_deref() {
             // We do not track indirect places.
             return;
         }

--- a/compiler/rustc_mir_dataflow/src/value_analysis.rs
+++ b/compiler/rustc_mir_dataflow/src/value_analysis.rs
@@ -34,6 +34,7 @@
 
 use std::collections::VecDeque;
 use std::fmt::{Debug, Formatter};
+use std::ops::Range;
 
 use rustc_data_structures::fx::FxHashMap;
 use rustc_index::bit_set::BitSet;
@@ -448,10 +449,8 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
 
     pub fn flood_with(&mut self, place: PlaceRef<'_>, map: &Map, value: V) {
         let StateData::Reachable(values) = &mut self.0 else { return };
-        map.for_each_aliasing_place(place, None, &mut |place| {
-            if let Some(vi) = map.places[place].value_index {
-                values[vi] = value.clone();
-            }
+        map.for_each_aliasing_place(place, None, &mut |vi| {
+            values[vi] = value.clone();
         });
     }
 
@@ -461,10 +460,8 @@ impl<V: Clone + HasTop + HasBottom> State<V> {
 
     pub fn flood_discr_with(&mut self, place: PlaceRef<'_>, map: &Map, value: V) {
         let StateData::Reachable(values) = &mut self.0 else { return };
-        map.for_each_aliasing_place(place, Some(TrackElem::Discriminant), &mut |place| {
-            if let Some(vi) = map.places[place].value_index {
-                values[vi] = value.clone();
-            }
+        map.for_each_aliasing_place(place, Some(TrackElem::Discriminant), &mut |vi| {
+            values[vi] = value.clone();
         });
     }
 
@@ -589,6 +586,9 @@ pub struct Map {
     projections: FxHashMap<(PlaceIndex, TrackElem), PlaceIndex>,
     places: IndexVec<PlaceIndex, PlaceInfo>,
     value_count: usize,
+    // The Range corresponds to a slice into `inner_values_buffer`.
+    inner_values: IndexVec<PlaceIndex, Range<usize>>,
+    inner_values_buffer: Vec<ValueIndex>,
 }
 
 impl Map {
@@ -598,6 +598,8 @@ impl Map {
             projections: FxHashMap::default(),
             places: IndexVec::new(),
             value_count: 0,
+            inner_values: IndexVec::new(),
+            inner_values_buffer: Vec::new(),
         }
     }
 
@@ -665,6 +667,14 @@ impl Map {
             // And push the eventual children places to the worklist.
             self.register_children(tcx, place, ty, &filter, &mut worklist);
         }
+
+        self.inner_values_buffer = Vec::with_capacity(self.value_count);
+        self.inner_values = IndexVec::from_elem(0..0, &self.places);
+        for local in body.local_decls.indices() {
+            if let Some(place) = self.locals[local] {
+                self.cache_preorder_invoke(place);
+            }
+        }
     }
 
     /// Potentially register the (local, projection) place and its fields, recursively.
@@ -718,6 +728,25 @@ impl Map {
         });
     }
 
+    /// Precompute the list of values inside `root` and store it inside
+    /// as a slice within `inner_values_buffer`.
+    fn cache_preorder_invoke(&mut self, root: PlaceIndex) {
+        let start = self.inner_values_buffer.len();
+        if let Some(vi) = self.places[root].value_index {
+            self.inner_values_buffer.push(vi);
+        }
+
+        // We manually iterate instead of using `children` as we need to mutate `self`.
+        let mut next_child = self.places[root].first_child;
+        while let Some(child) = next_child {
+            self.cache_preorder_invoke(child);
+            next_child = self.places[child].next_sibling;
+        }
+
+        let end = self.inner_values_buffer.len();
+        self.inner_values[root] = start..end;
+    }
+
     /// Returns the number of tracked places, i.e., those for which a value can be stored.
     pub fn tracked_places(&self) -> usize {
         self.value_count
@@ -768,11 +797,11 @@ impl Map {
     ///
     /// `tail_elem` allows to support discriminants that are not a place in MIR, but that we track
     /// as such.
-    pub fn for_each_aliasing_place(
+    fn for_each_aliasing_place(
         &self,
         place: PlaceRef<'_>,
         tail_elem: Option<TrackElem>,
-        f: &mut impl FnMut(PlaceIndex),
+        f: &mut impl FnMut(ValueIndex),
     ) {
         if place.is_indirect() {
             // We do not track indirect places.
@@ -789,7 +818,9 @@ impl Map {
             .chain(tail_elem.map(Ok).into_iter());
         for elem in elems {
             // A field aliases the parent place.
-            f(index);
+            if let Some(vi) = self.places[index].value_index {
+                f(vi);
+            }
 
             let Ok(elem) = elem else { return };
             let sub = self.apply(index, elem);
@@ -803,7 +834,7 @@ impl Map {
                 return;
             }
         }
-        self.preorder_invoke(index, f);
+        self.for_each_value_inside(index, f);
     }
 
     /// Invoke the given function on all the descendants of the given place, except one branch.
@@ -811,7 +842,7 @@ impl Map {
         &self,
         parent: PlaceIndex,
         preserved_child: Option<PlaceIndex>,
-        f: &mut impl FnMut(PlaceIndex),
+        f: &mut impl FnMut(ValueIndex),
     ) {
         for sibling in self.children(parent) {
             let elem = self.places[sibling].proj_elem;
@@ -821,16 +852,17 @@ impl Map {
                 // Only invalidate the other variants, the current one is fine.
                 && Some(sibling) != preserved_child
             {
-                self.preorder_invoke(sibling, f);
+                self.for_each_value_inside(sibling, f);
             }
         }
     }
 
-    /// Invoke a function on the given place and all descendants.
-    fn preorder_invoke(&self, root: PlaceIndex, f: &mut impl FnMut(PlaceIndex)) {
-        f(root);
-        for child in self.children(root) {
-            self.preorder_invoke(child, f);
+    /// Invoke a function on each value in the given place and all descendants.
+    fn for_each_value_inside(&self, root: PlaceIndex, f: &mut impl FnMut(ValueIndex)) {
+        let range = self.inner_values[root].clone();
+        let values = &self.inner_values_buffer[range];
+        for &v in values {
+            f(v)
         }
     }
 }

--- a/compiler/rustc_mir_dataflow/src/value_analysis.rs
+++ b/compiler/rustc_mir_dataflow/src/value_analysis.rs
@@ -632,11 +632,11 @@ impl Map {
         tcx: TyCtxt<'tcx>,
         body: &Body<'tcx>,
         filter: impl Fn(Ty<'tcx>) -> bool,
-        place_limit: Option<usize>,
+        value_limit: Option<usize>,
     ) -> Self {
         let mut map = Self::new();
         let exclude = excluded_locals(body);
-        map.register_with_filter(tcx, body, filter, exclude, place_limit);
+        map.register_with_filter(tcx, body, filter, exclude, value_limit);
         debug!("registered {} places ({} nodes in total)", map.value_count, map.places.len());
         map
     }
@@ -648,10 +648,11 @@ impl Map {
         body: &Body<'tcx>,
         filter: impl Fn(Ty<'tcx>) -> bool,
         exclude: BitSet<Local>,
-        place_limit: Option<usize>,
+        value_limit: Option<usize>,
     ) {
-        // We use this vector as stack, pushing and popping projections.
-        let mut worklist = VecDeque::with_capacity(place_limit.unwrap_or(body.local_decls.len()));
+        let mut worklist = VecDeque::with_capacity(value_limit.unwrap_or(body.local_decls.len()));
+
+        // Start by constructing the places for each bare local.
         self.locals = IndexVec::from_elem(None, &body.local_decls);
         for (local, decl) in body.local_decls.iter_enumerated() {
             if exclude.contains(local) {
@@ -668,8 +669,10 @@ impl Map {
         }
 
         // `place.elem1.elem2` with type `ty`.
+        // `elem1` is either `Some(Variant(i))` or `None`.
         while let Some((mut place, elem1, elem2, ty)) = worklist.pop_front() {
-            if let Some(place_limit) = place_limit && self.value_count >= place_limit {
+            // The user requires a bound on the number of created values.
+            if let Some(value_limit) = value_limit && self.value_count >= value_limit {
                 break
             }
 
@@ -688,6 +691,9 @@ impl Map {
             self.register_children(tcx, place, ty, &filter, &mut worklist);
         }
 
+        // Pre-compute the tree of ValueIndex nested in each PlaceIndex.
+        // `inner_values_buffer[inner_values[place]]` is the set of all the values
+        // reachable by projecting `place`.
         self.inner_values_buffer = Vec::with_capacity(self.value_count);
         self.inner_values = IndexVec::from_elem(0..0, &self.places);
         for local in body.local_decls.indices() {

